### DISCUSSION TITLE
Questions about DPA2 zero shot test

### DISCUSSION THREAD

**Discussion by duduRUHappy** [deepmodeling/deepmd-kit#3606] [Q&A] [answered]
When using the data and pre-trained Model.pt file publicly available in Zenodo (https://zenodo.org/records/10483908) for downstream testing tasks, there is a significant numerical deviation in the Energy RMSE values across multiple datasets, while the Force prediction values are largely consistent with those in the paper. Taking the H2O-PBE0TS-MD task as an example, my code is as follows:
dp_pt test -m /home/***/dpa2/Code_model_script/Model/DPA2_18head_1m/model.pt -f test_v1.json --head H2O_H2O-PD -d detail -n 2000

Below is the result log I obtained:

2024-03-26 11:08:12,891  [inference.py:223] INFO # ----------weighted average of errors----------- 
2024-03-26 11:08:12,891  [inference.py:227] INFO # number of systems for H2O_H2O-PD: 1
2024-03-26 11:08:12,891  [inference.py:229] INFO mae_e: 150.5802
2024-03-26 11:08:12,891  [inference.py:229] INFO mae_e_all: 28911.3991
2024-03-26 11:08:12,891  [inference.py:229] INFO mae_f: 0.0347
2024-03-26 11:08:12,891  [inference.py:229] INFO mae_v: 0.0286
2024-03-26 11:08:12,891  [inference.py:229] INFO rmse_e: 150.5802
2024-03-26 11:08:12,891  [inference.py:229] INFO rmse_f: 0.0444
2024-03-26 11:08:12,891  [inference.py:229] INFO rmse_v: 0.0391
2024-03-26 11:08:12,891  [inference.py:230] INFO # ----------------------------------------------- 

Why is this happening? If you need a more detailed log file, you can refer to the attached document.
[H2O.log](https://github.com/deepmodeling/deepmd-kit/files/14754511/H2O.log)

**Comment by iProzd** [comment, marked as answer]
When conducting a zero-shot test, it is important to adjust the energy bias due to the differences in DFT settings between the pretraining and downstream datasets. Failing to do so will result in a significant energy shift. To automate this process, you can set `numb_steps=1` (or maybe 0 is also ok) in the `input.json` file (containing downstream data systems, such as H2O-PBE0TS-MD,  placed in `training/training_data`) and perform a 1-step fine-tuning to update the energy bias based on the test dataset:

```
dp_pt train input.json --finetune model.pt --model-branch H2O_H2O-PD
```

This will generate a 1-step fine-tuned model (`model-1.pt`) with an updated energy bias, which you can then use for zero-shot testing.

Hope this will help you. @duduRUHappy

**Reply by duduRUHappy** [reply]
According to your advice, I implemented a 1-step fine-tune to obtain the model-1.pt, and then used it for zero-shot test. 
`dp_pt train /bohr/finetune-00sh/v1/input_mtl_finetue_H2O-PBE0TS-MD_1.json --finetune /bohr/Dpa2PretrainedModel-afjy/v1/Model/DPA2_18head_1m/model.pt --model-branch H2O_H2O-PD`
`dp_pt test -m model_1.pt -n 20000 -f /bohr/H20-H2O-PBE0TS-MD-D-p1mz/v3/boh_sysdir_test.json --head H2O_H2O-PD`

The rmse_e (0.1796) is somewhat better than before (150.5802), but there is still a significant gap compared to the results (rmse_e = 0.0006) in the paper.  Besides that, the rmse_f (0.6234) is much worse than before (0.0444). By the way,  the standard deviations of energy and force labels in the test set are 0.0033 and 0.8161.   And the json file used for finetune and the finetune result log file are attached. 
[input_mtl_finetue_H2O-PBE0TS-MD_1.json](https://github.com/deepmodeling/deepmd-kit/files/14785538/input_mtl_finetue_H2O-PBE0TS-MD_1.json)
[1-step_finetune_H2O-PBE0TS-MD.log](https://github.com/deepmodeling/deepmd-kit/files/14786031/1-step_finetune_H2O-PBE0TS-MD.log)


2024-03-27 16:20:27,811  [inference.py:211] INFO # ----------weighted average of errors----------- 
2024-03-27 16:20:27,811  [inference.py:213] INFO # number of systems : 1
2024-03-27 16:20:27,811  [inference.py:217] INFO mae_e: 0.1796
2024-03-27 16:20:27,811  [inference.py:217] INFO mae_e_all: 34.4790
2024-03-27 16:20:27,811  [inference.py:217] INFO mae_f: 0.4934
2024-03-27 16:20:27,812  [inference.py:217] INFO mae_v: 0.1078
2024-03-27 16:20:27,812  [inference.py:217] INFO rmse_e: 0.1796
2024-03-27 16:20:27,812  [inference.py:217] INFO rmse_f: 0.6234
2024-03-27 16:20:27,812  [inference.py:217] INFO rmse_v: 0.1636
2024-03-27 16:20:27,812  [inference.py:218] INFO # -----------------------------------------------

**Reply by zjgemi** [reply]
Yes, for the default initial learning rate 0.0002, even 1 step is too much to retain the accuracy. In practice, we use a sufficiently small learning rate (e.g. start_lr=stop_lr=1e-10) to finetune 1 step. Perhaps we should support 0 step to avoid this @iProzd. Currently 0 step will raise an error that numb_steps must be larger than warmup_steps.

**Comment by rajnichahal** [comment]
Does testing dpa2 model require deepmd v3 or we can do it using previous deepmd version? 
I tried testing with v2 and it is giving me the following error message:
Google.protobuf.message.DecodeError:Error parsing message

Please let me know. Thanks!

**Reply by duduRUHappy** [reply]
I tested DPA2 using the source code provided in the DPA2 paper (https://zenodo.org/records/10483908).

**Reply by rajnichahal** [reply]
Hi, Can you please share an example of how you proceeded with testing the model with this source code?

**Reply by duduRUHappy** [reply]
Maybe you can get some help from this notebook https://bohrium.dp.tech/notebooks/49345212168

**Reply by rajnichahal** [reply]
Thanks. I am looking into it.
Did you build this source code with tensorflow backend or with pytorch backend?

It shows me that model extension can be either .pb or .pth and I see your model's extension is .pt
Please see below:

-m MODEL, --model MODEL
Frozen model file (prefix) to import. TensorFlow backend: suffix is .pb; PyTorch backend: suffix is .pth. (default: frozen_model)

**Reply by rajnichahal** [reply]
Also, is it possible to load your own test dataset for testing using the dpa2 checkpoint on this notebook?
Thanks a lot!

**Reply by duduRUHappy** [reply]
If you wanna build the source code (https://bohrium.dp.tech/notebooks/49345212168), you can refer to the readme file for specific steps and requirements. Or you can just try the pre-release 3.0 version（https://github.com/deepmodeling/deepmd-kit/releases）. By the way, I'm also the new user of this platform, if you have more question, plz ask help from the tech team. Tks.
